### PR TITLE
pgbouncer: systemd support

### PR DESCRIPTION
**Pull request by 999eagle**
## Description of changes

This enables systemd support for pgbouncer itself and changes the service definition accordingly. See also the upstream example service file: https://github.com/pgbouncer/pgbouncer/blob/e6ce619785c93392794976c1e936e8c9f589a5ad/etc/pgbouncer.service.

## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/servers/sql/pgbouncer/default.nix [hunk 1]
```diff
@@ -10,8 +10,9 @@ stdenv.mkDerivation rec {
   };
 
   nativeBuildInputs = [ pkg-config ];
-  buildInputs = [ libevent openssl c-ares ];
+  buildInputs = [ libevent openssl c-ares systemd ];
   enableParallelBuilding = true;
+  configureFlags = [ "--with-systemd" ];
```
Comment (Scrumplex) [inline pkgs/servers/sql/pgbouncer/default.nix:15]: As this package seems to be cross-platform (`platforms.all`), this should probably be gated behind `stdenv.isLinux`

Comment (999eagle) [reply pkgs/servers/sql/pgbouncer/default.nix]: Fixed

**Comment by Scrumplex** [comment]
Result of `nixpkgs-review pr 285866` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>pgbouncer</li>
  </ul>
</details>

File: nixos/modules/services/databases/pgbouncer.nix [hunk 2]
```diff
@@ -96,10 +93,9 @@ in {
 
     logFile = mkOption {
       type = types.nullOr types.str;
-      default = "pgbouncer.log";
+      default = null;
       description = lib.mdDoc ''
         Specifies the log file.
```
Comment (Scrumplex) [inline nixos/modules/services/databases/pgbouncer.nix:98]: I think this should be a bit more precise

```suggestion
        Specified a log file in addition to journald.
```

File: nixos/modules/services/databases/pgbouncer.nix [hunk 3]
```diff
@@ -96,10 +93,9 @@ in {
 
     logFile = mkOption {
       type = types.nullOr types.str;
-      default = "pgbouncer.log";
+      default = null;
```
Comment (Scrumplex) [inline nixos/modules/services/databases/pgbouncer.nix:96]: This might be an unexpected change. Maybe add it to release 24.05 news?

**Comment by Scrumplex** [comment]
NixOS Test `pgbouncer` passes so apart from my review above, this LGTM

File: nixos/modules/services/databases/pgbouncer.nix [hunk 4]
```diff
@@ -601,22 +597,21 @@ in {
...
+      wants    = [ "network-online.target" ] ++ lib.optional config.services.postgresql.enable "postgresql.service";
+      after    = [ "network-online.target" ] ++ lib.optional config.services.postgresql.enable "postgresql.service";
       wantedBy = [ "multi-user.target" ];
       serviceConfig = {
-        Type = "forking";
+        Type = "notify";
         User = cfg.user;
         Group = cfg.group;
-        ExecStart = "${pkgs.pgbouncer}/bin/pgbouncer -d ${confFile}";
+        ExecStart = "${pkgs.pgbouncer}/bin/pgbouncer ${confFile}";
```
Comment (SuperSandro2000) [inline nixos/modules/services/databases/pgbouncer.nix:607]: We could maybe use lib.getExe here

File: pkgs/servers/sql/pgbouncer/default.nix [hunk 5]
```diff
@@ -10,8 +10,11 @@ stdenv.mkDerivation rec {
   };
 
   nativeBuildInputs = [ pkg-config ];
-  buildInputs = [ libevent openssl c-ares ];
+  buildInputs =
+    [ libevent openssl c-ares ]
```
Comment (SuperSandro2000) [inline pkgs/servers/sql/pgbouncer/default.nix:14]:
```suggestion
  buildInputs = [ libevent openssl c-ares ]
```

File: nixos/modules/services/databases/pgbouncer.nix (hunk 4 above)
Comment (999eagle) [reply nixos/modules/services/databases/pgbouncer.nix]: I feel like a lot of modules aren't using `lib.getExe` (yet?) but yeah it looks cleaner.

**Review by SuperSandro2000 — APPROVED** [review]

**Review by Scrumplex — APPROVED** [review]

**Pull request merged**